### ISSUE TITLE
Some issues with serverless deploy

### ISSUE THREAD

**Issue by rdaigle007** [ilsilfverskiold/ai-tech-news-bot#1]
Hi Ida!

I left a comment on your Medium article with some questions/issues.

After more issues doing "serverless deploy",  I thought maybe better to create an issue here to show the issues I had deploying to AWS,  and how I fixed them.  I think there was also an issue I mentioned in your medium article not referenced here that I also fixed in my environment.

I now have this setup in AWS serverless.  Looking forward to my first newsletter tomorrow morning.  :))))

Thanks so much for sharing this with the community.  It's pretty cool!

--------------------------------------------------------------

FYI, after peeling "serverless deploy" issues onion more, note I found the following changes also needed:

I'm super new to AWS serverless, etc.  So these aren't great solutions.  But FYI what I did to unblock myself.

1. Added following in serverless.yml for function newsletterTrigger:
  role: arn:aws:iam::801843946498:role/tech-bot-dev-us-east-1-lambdaRole

I'm guessing the first "serverless deploy" created relevant IAM role, and then subsequent deploys tried to create it also but already existed.

So this isn't a general solution, as this won't work the first time since the role isn't created yet.  I guess the "iamRoleStatements" in the serverless.yml causes the IAM role to be created 1st time, but fails subsequent times?

Error I was getting that this fixed was:

Error:
CREATE_FAILED: IamRoleLambdaExecution (AWS::IAM::Role)
Resource handler returned message: "Resource of type 'AWS::IAM::Role' with identifier 'tech-bot-dev-us-east-1-lambdaRole' already exists." (RequestToken: f0ea4c8b-fab9-c99b-1ef7-c31a9df78d87, HandlerErrorCode: AlreadyExists)

2. Then started getting another serverless deploy issue related to no permissions for events.

CREATE_FAILED: TechbotdevnewsletterTriggerrule2EventBridgeRule (AWS::Events::Rule)
Resource handler returned message: "User: arn:aws:iam::801843946498:user/serverless-newsbot is not authorized to perform: events:DescribeRule on resource: arn:aws:events:us-east-1:801843946498:rule/tech-bot-dev-newsletterTrigger-rule-2 because no identity-based policy allows the events:DescribeRule action (Service: EventBridge, Status Code: 400, Request ID: 747c5a19-5e08-4ebf-b677-df9310fdea35)" (RequestToken: 1b1b4ec8-e05f-a0b2-657c-8198bf2de27d, HandlerErrorCode: AccessDenied)

I I just added "events:*" to the serverless IAM policy you had us create.  That is probably too permissive, but... I needed to get things running.  :)

3. New issue with the schedule:

CREATE_FAILED: NewsletterTriggerEventsRuleSchedule1 (AWS::Events::Rule)
Resource handler returned message: "Parameter ScheduleExpression is not valid. (Service: EventBridge, Status Code: 400, Request ID: 5a5572f5-9f48-4eca-8dbc-c16e8c9023e4)" (RequestToken: 3f7735eb-3c1e-4941-f708-d9f10a639c6d, HandlerErrorCode: GeneralServiceException)

I changed the syntax of the schedules in serverless.yml to be like:

    events:
      - eventBridge:
          schedule: 'cron(0 13 ? * MON-FRI *)' # Trigger at 1PM UTC on weekdays
          input:
            time_period: "daily"
      - eventBridge:
          schedule: 'cron(0 13 ? * SAT *)' # Trigger at 1PM  UTC every Saturday
          input:
            time_period: "weekly"

**Comment by ilsilfverskiold** [comment]
rdaigle007, I seem to have excluded a few important bits.

As you recognized the issues are as follows (and have been fixed):

1. There was a typo in the serverless.yml file for the EventBridge
scheduled triggers. Please check that your serverless.yml file
corresponds to the new file I've pushed to this repo.

2. You did have to have Docker installed but should be able to package your
requirements without it. In this case change the serverless.yml file to
false under  dockerizePip: true.

3. There was a dependency missing called pydantic , please be sure to set
this in the requirements.txt file (see the repo as it is now - this should
work). If you don't set this there might be issues with your newsletter
going out at 10 AM UTC.

4. The python runtime was set as 3.8 where it should have been set to 3.11
or 3.12 (depending on your python version you had installed locally).

5. You did need a few more permissions for your IAM role that I hadn't pointed out, 
this should now have been added in via the readme file as well as this article 
so the next person don't end up with the same issues.

As you've managed to already deploy which is great, please just make sure
you have:

1. pydantic set in your requirements.txt file, you should be able to add it
and then simply run (if you have docker installed and set to true). If not,
please also run pip install -r requirements.txt and then severless deploy.

2. The serverless.yml file has the correct structure for the
eventBridge events.

Try this and then let me know if there are still issues.

**Comment by rdaigle007** [comment]
Thanks for the fast response!

I've made the changes you mentioned.  You were right, I did not get todays daily email, sounds like because of missing pydantic.  Looking forward to it tomorrow.

Purely FYI,  I still had "serverless deploy" issues if I didn't include the following 2 changes in my serverless.yml:
  - changed reservedConcurrency from 5 to 0.  (I still get that error when value is 5)
  - for some reason I still need to include the "role: arn:aws:iam::801843946498:role/tech-bot-dev-us-east-1-lambdaRole"

BTW, I did also updated my IAM users policy to include all the permissions you list in your readme.md.

The above 2 issues may just be some weird artifacts of my AWS account.   I'm happy as a clam, and looking forward to the report in the morning.   Then playing more over time.

Kind regards,
-Russ

**Comment by ilsilfverskiold** [comment]
Hmm, I do believe that setting the reservedConcurrency to 0 would essentially tell AWS Lambda that you do not want any instances of this function to run concurrently. This means that the Lambda function will not be able to process any events or triggers.

I'm unsure why you can't have 5 in concurrent executions set when deploying but essentially you can remove that one entirely then it would just default to account level limits. 

Can you leave it out or play with the numbers maybe set it to 1 or leave it out entirely?

You can also just change the concurrency in the AWS console as well. Under Configuration  -> Concurrency. Maybe this would be simpler.

Also, just good to just hit this locally (then at least it will send you the newsletter if you don't get it now at 10 AM UTC):
`serverless invoke local --function newsletterTrigger`

This is a bit trickier then it should be, but I promise it's always easier next time around.

**Comment by rdaigle007** [comment]
Yap, the eventBridge did not trigger function when configured with 0.   I commented out the reservedConcurrency in serverless.yml,   redeployed,  and the function was then triggered at the proper time and I received the email.    YAY!!!!

In case you are interested, I'll share a little findings below.  But I'm super happy with this working with it being disabled.

I played with the config in AWS console for the function, as you suggested.  Even setting reserved to 1 causes an error, because "The unreserved account concurrency can't go below 10".  And the function has a unreserved concurrency set to 10, so any reserved concurrency causes that to drop below 10.

Perhaps this is because of free-tier?  It seems any "provisioned" concurrency kicks you out of free tier for that lambda funciotn I think if I read the pricing docs for lambda properly.   

I'm a little confused at all these concurrency knobs.  For this newsbot, it doesn't seem critical to understand since only invoked once per day (well, possibly twice including the 1 day a week the weekly report is run).    Once I get to the point to have my OpenAI customGPT integrated with some serverless functions I'll be working on in the next month,  I'll probably have to understand this concurrency stuff since it'll have more requests coming (presumably via API gateway).

I'm happy to not only have your cool newsbot running, but also this is my first attempt to play with Lambda stuff... which I want to play with more.

Feel free to close this.  If you're curious to have me run some additional tests, I'm also happy to.

**Comment by rdaigle007** [comment]
Oops.... my mistake.  "provisioned concurrency" is different.     Apparently my account has pretty low unreserved concurrency.   I'm going to reach out to support to up the value.    Apparently depending when your AWS account was created,  there are different default values set onto accounts.

**Comment by ilsilfverskiold** [comment]
I find it strange that you can't set the reservedConcurrency, I think the default is much higher than 5 so technically it should be the other way around. Usually setting this is a way to control for traffic spikes, i.e. only 5 executions concurrently. 

I'm glad it worked though and you got introduce to working with lambda. :)
